### DISCUSSION TITLE
create your own adjacency matrix

### DISCUSSION THREAD

**Discussion by YunYunY** [pyg-team/pytorch_geometric#2948] [Q&A] [answered]
Hi,

I have two questions regarding how to batch multiple graphs with different number of nodes and create a new adjacency matrix

Background:\
To implement a new adjacency matrix like eq. 3 in this [paper](https://openaccess.thecvf.com/content_CVPR_2020/papers/Pan_Spatio-Temporal_Graph_for_Video_Captioning_With_Knowledge_Distillation_CVPR_2020_paper.pdf)

Two questions I have:
1. Batching multiple graphs (more than two) with different number of node in `get` method. Let's say I have 4 frames. This will give me 4 graphs with different node if I construct edge_index for each graph, but I also have 3 graphs with edge index between nodes in frame1 and frame2, frame2 and frame3, frame3 and frame4. This is the second type of graph. How could I pass 7 graphs (7 `Data` class) in `get` method of `torch_geometric.data.Dataset`\
There is a related [answer ](https://github.com/rusty1s/pytorch_geometric/issues/800) for this but I am not sure how to batch more than two graphs.

2.  According to [mini-batch documentation](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html), the adjacency matrices in one mini-batch is used to create a giant diagonal matrix. Suppose I could implement question 1 somehow, To implement a new adjacency matrix like the eq. 3, one method I thought was to modify the [source code of adj_t in gcn_norm](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/gcn_conv.html#GCNConv.forward). But how do I pass the information to distinguish between two types of edge_index? This means if `get` returns 4 type 1 edge_index and 3 type 2 edge_index, I want to put all type 1 related adjacency at [a11, a22, a33, a44], and put all type 2 related adjacency at [a12, a23, a34]. Is there a better method to do this? 

Thank you very much.

**Comment by rusty1s** [comment, marked as answer]
I think Eq. (3) in the paper is best implemented in a single `data` object. With this, you can simple store `G` in a single `edge_index` matrix where you increment the individual `edge_indices` manually, e.g., for the time dimension:
```python
for i in range(4):
    edge_indices[i][0] += ... # Add the number of previous nodes in the time dimension.
data.edge_index = torch.cat(edge_indices, dim=1)
```
Does that help you?

**Reply by YunYunY** [reply]
HI, 
Thank you for the quick response. This is so smart. I thought it in a very complicated way. Thank you!

**Reply by YunYunY** [reply]
Hi,
I have a follow up question after thinking about this method. If my understand is correct, the default graph is undirected. But what if part of them is directed but part of them is undirected. 

Say t_0 has three nodes [0, 1, 2], the connection is like 
```
edge_index_t1 = torch.tensor([[0, 1, 1, 2] ,[1, 0, 2, 1]], dtype=torch.long)
print(edge_index_t1)
tensor([[0, 1, 1, 2],
        [1, 0, 2, 1]])
```
This will result in a 3*3 adjacency matrix. 

If t_1 has two nodes, by adding the number of previous nodes in the time dimension, t_1 has two nodes with index [3, 4]. Suppose node 3 and 4 are connected, node 4 is connected with t_0's node 2 
```
edge_index_t12 = torch.tensor([[0, 1, 1, 2, 2, 4, 3, 4] ,[1, 0, 2, 1, 4, 2, 4, 3]], dtype=torch.long)
print(edge_index_t12)
tensor([[0, 1, 1, 2, 2, 4, 3, 4],
        [1, 0, 2, 1, 4, 2, 4, 3]])
```
This will result in 5*5 adjacency matrix A. A[:3] is exactly what I wanted. The right bottom corner A[3:, 3:] is also expressing the connection between node [3, 4] in t_1, but A[4, 2] has an element which is symmetric to A[2, 4]. We do need A[2, 4] that shows the connection between t_0 node 2 and t_1 node 4, but there is no connection between [4, 2] according to the eq. 3. How could I address this? Could I just remove the edge index [4,  2]?

**Reply by rusty1s** [reply]
Eq. (3) does not produce an undirected graph. If you want your graph to be undirected, you can simply convert it to an undirected one afterwards, e.g., via `torch_geometric.utils.to_undirected`. Hope that answers your question.

**Reply by YunYunY** [reply]
Thank you very much.
